File: linear-regression/linear_regression.cpp
#include <iostream>
#include <vector>
#include <numeric>
using namespace std;

class LinearRegression {
private:
    double w, b;
    
public:
    void fit(const vector<double>& X, const vector<double>& y) {
        double X_mean = accumulate(X.begin(), X.end(), 0.0) / X.size();
        double y_mean = accumulate(y.begin(), y.end(), 0.0) / y.size();
        
        double num = 0, den = 0;
        for (size_t i = 0; i < X.size(); i++) {
            num += (X[i] - X_mean) * (y[i] - y_mean);
            den += (X[i] - X_mean) * (X[i] - X_mean);
        }
        w = num / den;
        b = y_mean - w * X_mean;
    }
    
    double predict(double x) {
        return w * x + b;
    }
};

int main() {
    LinearRegression model;
    model.fit({1, 2, 3, 4, 5}, {2, 4, 6, 8, 10});
    cout << model.predict(6) << endl; // 12
}

File: linear-regression/linear_regression.rs
struct LinearRegression {
    w: f64,
    b: f64,
}

impl LinearRegression {
    fn new() -> Self {
        Self { w: 0.0, b: 0.0 }
    }
    
    fn fit(&mut self, x: &[f64], y: &[f64]) {
        let x_mean = x.iter().sum::<f64>() / x.len() as f64;
        let y_mean = y.iter().sum::<f64>() / y.len() as f64;
        
        let mut num = 0.0;
        let mut den = 0.0;
        for i in 0..x.len() {
            num += (x[i] - x_mean) * (y[i] - y_mean);
            den += (x[i] - x_mean).powi(2);
        }
        self.w = num / den;
        self.b = y_mean - self.w * x_mean;
    }
    
    fn predict(&self, x: f64) -> f64 {
        self.w * x + self.b
    }
}

fn main() {
    let mut model = LinearRegression::new();
    model.fit(&[1.0, 2.0, 3.0, 4.0, 5.0], &[2.0, 4.0, 6.0, 8.0, 10.0]);
    println!("{}", model.predict(6.0)); // 12
}
